### ISSUE TITLE
Your Training Set is Empty

### ISSUE THREAD

**Issue by mbittraining** [jantic/DeOldify#255]
Sadly, when running ImageColorizer.ipynb I get the following message on the third step:

C:\Users\Matthew\.conda\envs\deoldify\lib\site-packages\fastai\data_block.py:442: UserWarning: Your training set is empty. If this is by design, pass `ignore_empty=True` to remove this warning.
  warn("Your training set is empty. If this is by design, pass `ignore_empty=True` to remove this warning.")
C:\Users\Matthew\.conda\envs\deoldify\lib\site-packages\fastai\data_block.py:445: UserWarning: Your validation set is empty. If this is by design, use `split_none()`
                 or pass `ignore_empty=True` when labelling to remove this warning.
  or pass `ignore_empty=True` when labelling to remove this warning.""")

I am keen to see if I can download a default training set, or how to create a training set as at present I cannot use the product.

Please advise how I can overcome this issue?

Kind regards,

Matthew

**Comment by jantic** [comment]
@mbittraining There's two things going on here.  First that's clearly a Windows file path.  Windows isn't supported by us and you'll likely run into issues that we won't address here.

Second- that warning doesn't matter in terms of inference (just colorizing images).  It'll work regardless (given you don't have other issues running it on Windows).

I'll commit code to silence the warning shortly just so it doesn't cause further confusion though.

**Comment by mbittraining** [comment]
Hi Jason. I really appreciate the help. Can you please set up a Teams call
so that we can fix this together?

work account: matthew.Bennett@platformhg.com
personal:  mattybennett@outlook.com

Best regards,

Matthew
*Matthew Bennett*
*Freelance Trainer and IT Consultant*
*Mbe: **07871 480008*

<https://www.facebook.com/mattyjbennett>
<https://www.linkedin.com/in/matthew-bennett-a0384a11/>
<https://www.youtube.com/channel/UCwi6J8wbTkNQ4-wJpPS70Ug>
*[image: Microsoft Partner - Pinpoint page]*
* <http://www.fsb.org.uk/>*

**Comment by jantic** [comment]
I don't want to come across as rude, but no, this isn't something I'm
willing to do. This is being run on Windows which we don't support and
already stated that, and I just don't do free personal tech support like
that.

**Comment by leonoverweel** [comment]
@mbittraining have you tried running this in the cloud? Google Colab will let you run Python code on a GPU-enabled machine for free.

**Comment by mattybennett-phg** [comment]
A good idea, but from what I can make out, to be able to create your own training data you need to train using millions of images. The image bank Jason has used is absolutely massive.  (many tens of GB)

Speaking with others in the community using a limited number of images is also not the answer as the program with 'overtrain'. 

Not sure however how many would be 'reasonable'. 

If I wanted to stylise akin to Van Gogh for example I am limited to the number of paintings available in this style and colour scheme. 

However I am keen to learn if the pretrain data can be used as a starting point. Can you add existing colour images to a training bank thereby fixing the occasional problem? I've tried using the standard stable pretrained file and the results are very good, although it is very shy around the colour red. I have a person in the photo with a red handkerchief standing next to a blue wall. The wall turns out dark grey with a hint of blue, yet I have a colour original of this photo and know that it is a navy blue. 

The red handkerchief is a standard polka-dot silk handkerchief so should be a strong dark red. 

And as for tweed! I've got no chance there...

**Comment by mbittraining** [comment]
Hi Jason. I hope you are well.

Let's reset the conversation as I have an important question for you:

The training catalogue requires millions of images. This is not reasonable when you have a limited pallette and limited source images which are almost identical.

1. How do I limit the number of training images and get an optimal result when colourising only one image and have similar images taken at the same time (e.g. photoshoot of 10 identical images with the one b/w image to be colourised slightly offset)?

2. To help with the quantity can I mirror and flip the images so I have 4 rotations of each image? Would that help or does it matter at all? Does the algorithm boil down the colour neighbouring associations therefore it doesn't matter to rotate them?

I'd value your ideas on the above.

Best regards,

Matthew

**Comment by jantic** [comment]
@mbittraining Answers:

1. What you're really describing here is what [Deep Exemplar-based Colorization](https://github.com/msracver/Deep-Exemplar-based-Colorization) tackles best.  DeOldify would realistically require a lot more examples than 10, and even if you had the data it's questionable as to how well it would work relative to expectations.  Reason: DeOldify leans heavily on more generalized material/physical properties to determine color rather than object recognition.  This makes it robust but also limited in accuracy.
2. DeOldify already does this with the data you feed it and much more.  This is what is termed "augmentation" and is a standard practice to get more mileage out of your training data. It definitely helps but it won't be the magic bullet in your scenario here.

**Comment by mattybennett-phg** [comment]
Perfect. Thank you for your help with this, Jason. I'll look into Deep Exemplar now.
